src/parser/csql_grammar.y

```diff
@@ -5185,45 +5192,125 @@ only_all_class_spec
 		DBG_PRINT}}
 	;
 
-class_name
-	: identifier DOT identifier
+user_specified_name_without_dot
+	: identifier_without_dot DOT identifier_without_dot
 		{{
 
-			PT_NODE *user_node = $1;
-			PT_NODE *name_node = $3;
+			PT_NODE *user = $1;
+			PT_NODE *name = $3;
 
-			if (name_node != NULL && user_node != NULL)
+			if (user
+			    && name
+			    && user->node_type == PT_NAME
+			    && name->node_type == PT_NAME)
 			  {
-			    name_node->info.name.resolved = pt_append_string (this_parser, NULL,
-			                                                      user_node->info.name.original);
+			    name->info.name.resolved = pt_append_string (this_parser, NULL, user->info.name.original);
+
+			    parser_free_tree (this_parser, user);
+
+			    // PT_NAME_INFO_SET_FLAG (name, PT_NAME_INFO_USER_SPECIFIED);
+			  }
+
+			$$ = name;
+			PARSER_SAVE_ERR_CONTEXT ($$, @$.buffer_pos)
+
+		DBG_PRINT}}
+	| identifier_without_dot
+		{{
+
+			PT_NODE *name = $1;
+
+			if (name
+			    && name->node_type == PT_NAME)
+			  {
+			    // PT_NAME_INFO_SET_FLAG (name, PT_NAME_INFO_USER_SPECIFIED);
 			  }
-			if (user_node != NULL)
+
+			$$ = name;
+			PARSER_SAVE_ERR_CONTEXT ($$, @$.buffer_pos)
+
+		DBG_PRINT}}
+	;
+
+user_specified_name
+	: identifier DOT identifier
+		{{
+
+			PT_NODE *user = $1;
+			PT_NODE *name = $3;
+
+			if (user
+			    && name
+			    && user->node_type == PT_NAME
+			    && name->node_type == PT_NAME)
 			  {
-			    parser_free_tree (this_parser, user_node);
+			    name->info.name.resolved = pt_append_string (this_parser, NULL, user->info.name.original);
+
+			    parser_free_tree (this_parser, user);
+			    
+			    // PT_NAME_INFO_SET_FLAG (name, PT_NAME_INFO_USER_SPECIFIED);
 			  }
 
-			$$ = name_node;
+			$$ = name;
 			PARSER_SAVE_ERR_CONTEXT ($$, @$.buffer_pos)
 
 		DBG_PRINT}}
 	| identifier
 		{{
 
-			$$ = $1;
+			PT_NODE *name = $1;
+
+			if (name
+			    && name->node_type == PT_NAME)
```

Comment:
Isn't there a case where it is an identifier but not PT_NAME?

Reply:
Okay. I'll just check for NULL.